[Merged by Bors] - feat(Probability/Independence/Basic): upstream `iIndepFun_iff_pi_map_eq_map` from PFR

hanwenzhu opened the pull request:
* Added lemma `aemeasurable_pi_lambda` for `AEMeasurable (f : α → ∀ a : δ, X a) μ` for countable `δ`, parallel to `measurable_pi_lambda`.
* Showed that a finite collection of random variables are independent iff their joint law is the product distribution of their respective laws.

This lemma is needed in CLT.

I am using git blame on PFR to figure out the person (@MantasBaksys) who wrote this and adding to co-authored-by field:

Co-authored-by: Mantas Bakšys <baksys.mantas@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RemyDegenne commented on Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:
```diff
@@ -419,4 +419,21 @@ theorem Measurable.comp_aemeasurable' [MeasurableSpace δ] {f : α → δ} {g :
+variable {δ : Type*} [Countable δ] {X : δ → Type*} [∀ a, MeasurableSpace (X a)]
+
+theorem aemeasurable_pi_iff {g : α → ∀ a, X a} :
+    AEMeasurable g μ ↔ ∀ a, AEMeasurable (fun x => g x a) μ := by
```
The new code is mixing `=>` and `↦` for functions. Let's use only one.
```suggestion
    AEMeasurable g μ ↔ ∀ a, AEMeasurable (fun x ↦ g x a) μ := by
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:
```diff
@@ -419,4 +419,21 @@ theorem Measurable.comp_aemeasurable' [MeasurableSpace δ] {f : α → δ} {g :
 
+variable {δ : Type*} [Countable δ] {X : δ → Type*} [∀ a, MeasurableSpace (X a)]
+
+theorem aemeasurable_pi_iff {g : α → ∀ a, X a} :
```
I find the use of forall a bit strange for a product type.
```suggestion
theorem aemeasurable_pi_iff {g : α → Π a, X a} :
```

hanwenzhu replied:
Thanks, I definitely agree. I copied signature from `measurable_pi_iff`. Should I also change this notation there?

RemyDegenne commented on Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:
```diff
@@ -419,4 +419,21 @@ theorem Measurable.comp_aemeasurable' [MeasurableSpace δ] {f : α → δ} {g :
+    exact (eventually_countable_forall.mpr fun a ↦ (h a).ae_eq_mk).mono fun _ h ↦ funext h
+
+@[fun_prop, aesop safe 100 apply (rule_sets := [Measurable])]
+theorem aemeasurable_pi_lambda (f : α → ∀ a, X a) (hf : ∀ a, AEMeasurable (fun c => f c a) μ) :
```
```suggestion
theorem aemeasurable_pi_lambda (f : α → Π a, X a) (hf : ∀ a, AEMeasurable (fun c ↦ f c a) μ) :
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:
```diff
@@ -419,4 +419,21 @@ theorem Measurable.comp_aemeasurable' [MeasurableSpace δ] {f : α → δ} {g :
     (hg : Measurable g) (hf : AEMeasurable f μ) : AEMeasurable (fun x ↦ g (f x)) μ :=
   Measurable.comp_aemeasurable hg hf
 
+variable {δ : Type*} [Countable δ] {X : δ → Type*} [∀ a, MeasurableSpace (X a)]
```
```suggestion
variable {δ : Type*} [Countable δ] {X : δ → Type*} {mX : ∀ a, MeasurableSpace (X a)}
```
To allow for sigma-algebras that are not the one found by typeclass inference.

hanwenzhu replied:
Same as above, I copied this from `measurable_pi_iff` in `Mathlib/MeasureTheory/MeasurableSpace/Basic.lean`. Should I also change the notation in that file? (perhaps in separate PR)

RemyDegenne replied:
Another PR modifying that other file would be welcome. Let's not do this in this PR though.

RemyDegenne commented on Mathlib/Probability/Independence/Basic.lean:
```diff
@@ -731,6 +731,31 @@ lemma iIndepFun_precomp_of_bijective {g : ι' → ι} (hg : g.Bijective) :
   mp := .of_precomp hg.surjective
   mpr := .precomp hg.injective
 
+theorem iIndepFun_iff_pi_map_eq_map [Fintype ι] [IsProbabilityMeasure μ]
```
Can you move that next to `ProbabilityTheory.indepFun_iff_map_prod_eq_prod_map_map`?

RemyDegenne commented on Mathlib/Probability/Independence/Basic.lean:
```diff
@@ -731,6 +731,31 @@ lemma iIndepFun_precomp_of_bijective {g : ι' → ι} (hg : g.Bijective) :
   mp := .of_precomp hg.surjective
   mpr := .precomp hg.injective
 
+theorem iIndepFun_iff_pi_map_eq_map [Fintype ι] [IsProbabilityMeasure μ]
```
`indepFun_iff_map_prod_eq_prod_map_map` also works for finite measures. Can you also relax the `IsProbabilityMeasure μ` assumption here to have `IsFiniteMeasure μ` instead?

hanwenzhu replied:
I tried to do that but had limited success: I don't need any assumption for the forward direction because of `iIndepFun.isProbabilityMeasure`. But for the reverse direction there are special cases: if ι is singleton, then it seems the RHS of ↔ is unconditionally true, while LHS requires µ to be a probability measure.

RemyDegenne commented on Mathlib/Probability/Independence/Basic.lean:
```diff
@@ -731,6 +731,31 @@ lemma iIndepFun_precomp_of_bijective {g : ι' → ι} (hg : g.Bijective) :
   mpr := .precomp hg.injective
 
+theorem iIndepFun_iff_pi_map_eq_map [Fintype ι] [IsProbabilityMeasure μ]
+    (hf : ∀ (x : ι), AEMeasurable (f x) μ) :
```
```suggestion
    (hf : ∀ i, AEMeasurable (f i) μ) :
```
That variable is denoted by i everywhere in the proof.

RemyDegenne reviewed (commented):
Thanks for upstreaming this!

hanwenzhu commented:
One style question I had but forgot to ask is: currently `iIndepFun_iff_pi_map_eq_map` and `indepFun_iff_map_prod_eq_prod_map_map` have different directions (`pi_eq_map` vs `map_eq_prod`). I think they (and their naming) should be unified, but I don't know which direction makes more sense.

RemyDegenne commented:
The rewriting direction of `indepFun_iff_map_prod_eq_prod_map_map` seems the more natural to me: you use independence to say that the joint law is a product.

RemyDegenne commented on Mathlib/Probability/Independence/Basic.lean:
```diff
@@ -615,6 +615,32 @@ theorem indepFun_iff_map_prod_eq_prod_map_map {mβ : MeasurableSpace β} {mβ' :
+  · intro h S s hs
+    specialize h₀ (s := fun i ↦ if i ∈ S then s i else univ)
+      fun i ↦ by beta_reduce; split_ifs with hiS <;> simp [hiS, hs]
+    simp [← h, apply_ite, Finset.prod_ite, iInter_ite] at h₀
```
That's a non-terminal `simp`. Please replace it by a `simp only`.

hanwenzhu replied:
Nice catch, sorry I didn't notice that! Thank you!

hanwenzhu commented:
> The rewriting direction of `indepFun_iff_map_prod_eq_prod_map_map` seems the more natural to me: you use independence to say that the joint law is a product.

Thanks, I changed the direction of my newly added theorem. After changing the direction, I renamed to `iIndepFun_iff_map_fun_eq_pi_map` because in my opinion it makes more sense than `iIndepFun_iff_map_eq_pi_map`; but if you have any name suggestions I am very happy to change it

RemyDegenne commented:
Thanks!
bors r+
